### PR TITLE
Feature/average function

### PR DESCRIPTION
**Pull request by twennemann**
New Branche for average function.

Tests will come soon

### CODE REVIEW

File: pyfar/dsp/dsp.py [hunk 1]
```diff
@@ -1498,3 +1498,124 @@ def decibel(signal, domain='freq', log_prefix=None, log_reference=1,
...
+    -----
+    The functions :py:func:`~pyfar.dsp.linear_phase` and
+    :py:func:`~pyfar.dsp.minimum_phase` can be used to obtain non-zero phase
+    responses. This can be usefull if the average mode discards the phase and
+    the paramter `phase_copy` can not be used.
+    """
+
+    # check input
+    if not isinstance(signal, pyfar.Signal):
+        raise TypeError('Input data has to be of type: Signal')
```
Comment (f-brinkmann) [inline pyfar/dsp/dsp.py:1558]: can now use TimeData and FrequencyData depending on the mode

File: pyfar/dsp/dsp.py [hunk 2]
```diff
@@ -1498,3 +1498,124 @@ def decibel(signal, domain='freq', log_prefix=None, log_reference=1,
...
+
+    # set weights default
+    if weights is None:
+        weights = 1/(np.prod(signal.cshape))
+    else:
+        weights = weights/np.sum(weights)
```
Comment (f-brinkmann) [inline pyfar/dsp/dsp.py:1564]: If we allow weights that can be broadcasted to cshape, we should not do the normalization. And add a note to the docstring that weights usually sum to 1 for most applications. Instead we could test if they can be braodcasted and raise a ValueError if not...

File: pyfar/dsp/dsp.py [hunk 3]
```diff
@@ -1498,3 +1498,124 @@ def decibel(signal, domain='freq', log_prefix=None, log_reference=1,
...
+
+    # convert data to desired domain
+    if mode == 'time':
+        data = signal.time.copy()
+    elif mode == 'complex':
+        data = signal.freq.copy()
+    elif mode == 'magnitude':
+        data = np.abs(signal.freq.copy())
+    elif mode == 'power':
+        data = np.abs(signal.freq.copy())**2
```
Comment (f-brinkmann) [inline pyfar/dsp/dsp.py:1574]: copy() operation not needed - np.abs does it for us.

File: pyfar/dsp/dsp.py [hunk 4]
```diff
@@ -1498,3 +1498,124 @@ def decibel(signal, domain='freq', log_prefix=None, log_reference=1,
...
+                                             prefix_return=True)
+
+    # apply weights
+    # NOT SURE IF THIS WORKS WITH MORE THAN FOR SIGNALS GREATER THAN 3D
+    # data = data * ca._get_arithmetic_data(weights,None, None)
+    data = data * np.asarray(weights)
+
+    # average the data
+    if (mode == 'time' or
+            mode == 'complex' or mode == 'magnitude'):
```
Comment (f-brinkmann) [inline pyfar/dsp/dsp.py:1586]: shorter
```suggestion
    if mode in ['time', 'complex', 'magnitude']:
```

File: pyfar/dsp/dsp.py [hunk 5]
```diff
@@ -1498,3 +1498,124 @@ def decibel(signal, domain='freq', log_prefix=None, log_reference=1,
...
+        data = np.sum(data, axis=axis, keepdims=True)
```
Comment (f-brinkmann) [inline pyfar/dsp/dsp.py:1587]: - The sum is done in any case and could go before the if, else.
- I would suggest to make keepdims a parameter with the default being False. In that case, we would add ´np.squeeze(data, axis)` after the sum (where we keep keepdims)

File: pyfar/dsp/dsp.py [hunk 6]
```diff
@@ -1498,3 +1498,124 @@ def decibel(signal, domain='freq', log_prefix=None, log_reference=1,
         return log_prefix * np.log10(np.abs(data) / log_reference), log_prefix
     else:
         return log_prefix * np.log10(np.abs(data) / log_reference)
+
+
+def average(signal, mode='time', axis=None, phase_copy=None,
```
Comment (f-brinkmann) [inline pyfar/dsp/dsp.py:1503]: docstring and checks for axis are missing. We have to make sure that `np.max(axis) <= len(signal.cshape)`

**Review by f-brinkmann — CHANGES_REQUESTED** [review]
Thanks for opening this - found a few minor things and a few things to discuss in the weekly

File: pyfar/dsp/dsp.py [hunk 7]
```diff
@@ -1498,3 +1499,138 @@ def decibel(signal, domain='freq', log_prefix=None, log_reference=1,
...
+    if 1 in signal.cshape:
+        for i, ax in enumerate(axis):
+            if signal.cshape[ax] == 1:
+                warnings.warn("Sinnvolles Warning")
+    # set weights default
+    if weights is None:
+        weights = 1
+        for idx, size in enumerate(signal.cshape):
+            if idx in axis:
+                weights /= signal.cshape[idx]
```
Comment (mberz) [inline pyfar/dsp/dsp.py:1596]: Can't we do this without a loop?

File: pyfar/dsp/dsp.py [hunk 8]
```diff
@@ -1498,3 +1499,138 @@ def decibel(signal, domain='freq', log_prefix=None, log_reference=1,
...
+    # convert data to desired domain
+    if mode == 'time':
+        data = signal.time
+    elif mode == 'complex':
+        data = signal.freq
+    elif mode == 'magnitude_zerophase':
+        data = np.abs(signal.freq)
+    elif mode == 'magnitude_phase':
+        data = np.abs(signal.freq) * \
+               np.exp(1j*pyfar.dsp.phase(signal, unwrap=True))
```
Comment (mberz) [inline pyfar/dsp/dsp.py:1611]: This is the exact same as `mode='complex'`

File: pyfar/dsp/dsp.py [hunk 9]
```diff
@@ -1498,3 +1499,138 @@ def decibel(signal, domain='freq', log_prefix=None, log_reference=1,
...
+    elif mode == 'log_magnitude_zerophase':
+        data, log_prefix = pyfar.dsp.decibel(signal, 'freq',
+                                             prefix_return=True)
+    else:
+        raise ValueError(
+            """mode must be 'time', 'complex', 'magnitude_zerophase', 'power',
+            'magnitude_phase' or 'log_magnitude_zerophase'."""
+            )
+    # average the data
+    data = np.sum(data, axis=axis, keepdims=keepdims)
```
Comment (mberz) [inline pyfar/dsp/dsp.py:1623]: You can also use np.average, which supports weighted averaging. 
https://numpy.org/doc/stable/reference/generated/numpy.average.html
This way you can also avoid the looped version of calculating the weights.

**Review by mberz — CHANGES_REQUESTED** [review]
This PR is marked as requirement for the next release, but looks unfinished. 
What's the status on this? I'd suggest moving this to the next release.

**Review by f-brinkmann — APPROVED** [review]

**Review by mberz — CHANGES_REQUESTED** [review]
1. There's a couple of things in the docstring which require fixes.
2. Please don't copy the entire input signal only to create a new Signal, TimeData or Freq.Data

File: pyfar/dsp/dsp.py [hunk 10]
```diff
@@ -1851,6 +1851,143 @@ def rms(signal):
...
+        ``'log_magnitude_zerophase'``
+            average the logarithmic magnitude spectra $20 \\log_{10}(X)$ and
+            discard the phase. The logarithm is reversed before returning the
+            averaged signal.
+
+        The default is ``'time'``
+    axis: None, int, or tuple of ints, optional
+        Axis or axes along which the averaging is done. Can be None, which will
+        take average across all axes signal. Negative values refer to
+        ``signal.cshape`` to avoid averaging across the time or frequency axis.
```
[the remark this answers is not in the document]
Comment (f-brinkmann) [reply pyfar/dsp/dsp.py:1887]: Good point, this is too short at the moment. I would suggest to postpone this until we change the parameter name to `caxis`. That will add content to the concepts that could be well written and linked from here.

File: pyfar/dsp/dsp.py [hunk 11]
```diff
@@ -1851,6 +1851,138 @@ def rms(signal):
...
+    signal: Signal, TimeData, FrequencyData
+        Input signal as pyfar audio object.
+    mode: string
+
+        ``'linear'``
+            average ``signal.time`` if the signal is in the time domain and
+            ``signal.freq`` if the signal is in the frequency domain. Note that
+            these operations are equivalent due to the linearity of the
+            averaging and the FFT. This mode might cause artifacts if the
+            data is not aligned across channels.
```
Comment (sikersten) [inline pyfar/dsp/dsp.py:1869]:
```suggestion
            Average ``signal.time`` if the signal is in the time domain and
            ``signal.freq`` if the signal is in the frequency domain. Note that
            these operations are equivalent for `Signal` objects due to the
            linearity of the averaging and the FFT. This mode might cause
            artifacts if the data is not aligned across channels.
```

File: pyfar/dsp/dsp.py [hunk 12]
```diff
@@ -1851,6 +1851,138 @@ def rms(signal):
...
+        ``'magnitude_zerophase'``
+            average the magnitude spectra and discard the phase
+        ``'magnitude_phase'``
+            average the magnitude spectra and the unwrapped phase seperatly.
```
Comment (sikersten) [inline pyfar/dsp/dsp.py:1873]:
```suggestion
            Average the magnitude spectra and the unwrapped phase separately.
```

File: pyfar/dsp/dsp.py [hunk 13]
```diff
@@ -1851,6 +1851,138 @@ def rms(signal):
...
+        ``'power'``
+            average the power spectra :math:`|X|^2` and discard the phase. The
+            squaring of the spectra is reversed before returning the averaged
+            signal.
+        ``'log_magnitude_zerophase'``
+            average the logarithmic magnitude spectra
+            :math:`20 \\log_{10}(|X|)` and discard the phase. The logarithm is
+            reversed before returning the averaged signal.
```
Comment (sikersten) [inline pyfar/dsp/dsp.py:1881]:
```suggestion
            Average the logarithmic magnitude spectra
            using :py:func:`~pyfar.dsp.decibel` and discard the phase. The
            logarithm is reversed before returning the averaged signal.
```

File: pyfar/dsp/dsp.py [hunk 14]
```diff
@@ -1851,6 +1851,138 @@ def rms(signal):
...
+        channels specified in the tuple. The default is ``None``.
+    weights: array like
+        array with channel weights for averaging the data. Must be
+        broadcastable to ``signal.cshape``. The default is ``None``, which
+        applies equal weights to all channels.
+    keepdims: bool, optional
+        If this is ``True``, the axes which are reduced during the averaging
+        are kept as a dimension with size one. Otherwise, singular dimensions
+        will be squeezed after averaging. The default is ``False``.
+    Returns
```
Comment (sikersten) [inline pyfar/dsp/dsp.py:1898]:
```suggestion
    
    Returns
```

File: pyfar/dsp/dsp.py [hunk 15]
```diff
@@ -1851,6 +1851,138 @@ def rms(signal):
...
+    --------
+    averaged_signal: Signal, TimeData, FrequencyData
+        averaged input Signal
```
Comment (sikersten) [inline pyfar/dsp/dsp.py:1901]:
```suggestion
        Averaged input Signal.
```

File: pyfar/dsp/dsp.py [hunk 16]
```diff
@@ -1851,6 +1851,138 @@ def rms(signal):
...
+
+    Notes
+    -----
+    The functions :py:func:`~pyfar.dsp.linear_phase` and
+    :py:func:`~pyfar.dsp.minimum_phase` can be used to obtain non-zero phase
+    responses. This can be usefull if the average mode discards the phase.
+
```
Comment (sikersten) [inline pyfar/dsp/dsp.py:1908]:
```suggestion
    :py:func:`~pyfar.dsp.minimum_phase` can be used to obtain a phase for
    magnitude spectra after using a mode that discards the phase.
```

**Review by sikersten — APPROVED** [review]

**Pull request merged**